jekyll/_cci2/docker-layer-caching.md

```diff
@@ -0,0 +1,24 @@
+---
+layout: classic-docs
+title: "Docker Layer Caching"
+short-title: "Docker Layer Caching"
+order: 5
+---
+
+{% include beta-premium-feature.html%}
+
+If your application is distributed as a Docker image you are probably know that Docker image consists of layers and most of the time only top layers are changing. So if your image is a heavy one you may want to reuse not changed layers and significantly reduce time for building image. By default [Remote Docker Environment]({{ site.baseurl }}/2.0/building-docker-images) doesn't provide caching for layers, but it can be enabled with a special option:
+
+``` YAML
+- setup_docker_engine:
+    reusable: true    # default - false
+    exclusive: true   # default - true
+```
+
+Please note, that this feature requires your project to be whitelisted. Please reach out to your support representative in order to enable it.
+
+In case of `reusable: true` CircleCI will try to reuse your previous Remote Docker Environment so all the layers you built in previous job(s) will be there. We try to reuse previous environment at a best efforts basis. That means that in some cases your job may find itself in a clean environment even if your config specifies `reusable: true`.
+
+If you run many parallel jobs for the same project (the same or different branches) and each of them are requesting reusable environment none of them will be blocked and all will be provided with a Remove Docker Environment (but probably not all of them will have cached layers). Please note that this feature is in Beta right now and we might change this behaviour later.
+
+The second option `exclusive` indicates whether you want to allow running parallel jobs in the same Remote Docker Environment simultaneously or not. Disabling it (`exclusive: false`) will make parallel jobs of the same project share one Remote Docker Environment. This way you can reduce chances of getting a clean Remote Docker. However you need to be cautious and make sure that your project is tolerant to shared Docker environment. For instance, if you build images with a mutable tags (like `latest`) then shared Docker environment might be a bad idea. In case of `exclusive: false` we will allow only two parallel jobs to use the same Docker environment. Please note that this feature is in Beta right now and we might change this behaviour later.
```

Comment:
Quoting since this is merged already:

> If you specify exclusive: true, your job is guaranteed to have an exclusive Remote Docker Environment that other jobs can’t access.
>
> If you specify exclusive: false, a project’s parallel jobs will be able to share the same Remote Docker Environment. This method lets you reduce the chances of receiving a Remote Docker Environment without a cache. If you choose this option, ensure that your project can handle concurrent operations in Docker Engine.
> 
> For example, if you build images with mutable tags like latest, then a shared Docker environment could be a bad idea. With exclusive: false, we only allow 2 parallel jobs to use the same Docker environment.

Okay, this is clearer to me now.  To restate this and confirm I understand: if you use immutable tags, and you're certain your jobs can otherwise run in the same Docker environment as each other, you can use `exclusive: false` safely.  You'll benefit from a higher chance of cache-hits.

If you're unsure, you should use `exclusive: true`, as you risk running your jobs with inconsistent environments and creating hard-to-debug problems.  Is that about right?

Nice-to-have additional questions:
With `exclusive: false`, is every 2nd parallel job guaranteed to land on the machine of the 1st parallel job?  Does this apply to every 2N and 2N-1 job as well?